Hidden mypy errors

On our PR tester, we use `mypy-1.11.2-cp39` and the type checks pass. However, when I run `mypy-1.11.2-cp310` I see typing errors:

```
$ mypy -p src.tribler.core
src/tribler/core/restapi/webui_endpoint.py:40: error: Incompatible types in assignment (expression has type "FileResponse", variable has type "ClientResponse")  [assignment]
src/tribler/core/restapi/webui_endpoint.py:41: error: Property "content_type" defined in "HeadersMixin" is read-only  [misc]
src/tribler/core/restapi/webui_endpoint.py:41: error: Incompatible types in assignment (expression has type "str | None", variable has type "str")  [assignment]
src/tribler/core/restapi/rest_manager.py:264: error: Item "AbstractServer" of "AbstractServer | None" has no attribute "sockets"  [union-attr]
src/tribler/core/restapi/rest_manager.py:264: error: Item "None" of "AbstractServer | None" has no attribute "sockets"  [union-attr]
src/tribler/core/restapi/rest_manager.py:283: error: Item "AbstractServer" of "AbstractServer | None" has no attribute "sockets"  [union-attr]
src/tribler/core/restapi/rest_manager.py:283: error: Item "None" of "AbstractServer | None" has no attribute "sockets"  [union-attr]
src/tribler/core/restapi/file_endpoint.py:54: error: Incompatible types in assignment (expression has type "Path", variable has type "str")  [assignment]
src/tribler/core/restapi/file_endpoint.py:55: error: "str" has no attribute "is_dir"  [attr-defined]
src/tribler/core/restapi/file_endpoint.py:56: error: "str" has no attribute "parent"  [attr-defined]
src/tribler/core/restapi/file_endpoint.py:60: error: "str" has no attribute "iterdir"  [attr-defined]
src/tribler/core/restapi/file_endpoint.py:71: error: "str" has no attribute "parent"  [attr-defined]
src/tribler/core/restapi/file_endpoint.py:75: error: "str" has no attribute "resolve"  [attr-defined]
src/tribler/core/restapi/events_endpoint.py:257: error: "RequestHandler" has no attribute "_handler_waiter"  [attr-defined]
src/tribler/core/restapi/events_endpoint.py:258: error: "RequestHandler" has no attribute "_handler_waiter"  [attr-defined]
src/tribler/core/libtorrent/restapi/downloads_endpoint.py:411: error: Incompatible types in assignment (expression has type "int", target has type "str")  [assignment]
src/tribler/core/libtorrent/restapi/downloads_endpoint.py:413: error: Incompatible types in assignment (expression has type "bool", target has type "str")  [assignment]
src/tribler/core/libtorrent/restapi/downloads_endpoint.py:418: error: Incompatible types in assignment (expression has type "list[int]", target has type "str")  [assignment]
Found 18 errors in 5 files (checked 93 source files)
```

For some reason, our PR checks also scan 98 files instead of the 93 I get locally (`Success: no issues found in 98 source files`).

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -5,10 +5,14 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
+        with:
+          submodules: 'true'
       - name: Setup Python 3.9
         uses: actions/setup-python@v5
         with:
           python-version: 3.9
+          cache: 'pip'
+      - run: python -m pip install -r requirements.txt
       - name: Install mypy
         run: pip install mypy
       - name: Run mypy

--- a/src/tribler/core/libtorrent/restapi/downloads_endpoint.py
+++ b/src/tribler/core/libtorrent/restapi/downloads_endpoint.py
@@ -400,17 +400,20 @@ class DownloadsEndpoint(RESTEndpoint):
         "uri*": (String, "The URI of the torrent file that should be downloaded. This URI can either represent a file "
                          "location, a magnet link or a HTTP(S) url."),
     }))
-    async def add_download(self, request: Request) -> RESTResponse:
+    async def add_download(self, request: Request) -> RESTResponse:  # noqa: C901
         """
         Start a download from a provided URI.
         """
         tdef = uri = None
         if request.content_type == 'applications/x-bittorrent':
-            params = dict(request.query.items())
-            if 'anon_hops' in params:
-                params['anon_hops'] = int(params['anon_hops'])
-            if 'safe_seeding' in params:
-                params['safe_seeding'] = params['safe_seeding'] != 'false'
+            params: dict[str, str | int | list[int]] = {}
+            for k, v in request.query.items():
+                if k == "anon_hops":
+                    params[k] = int(v)
+                elif k == "safe_seeding":
+                    params[k] = v != "false"
+                else:
+                    params[k] = v
             body = await request.read()
             metainfo = cast(dict[bytes, Any], lt.bdecode(body))
             packed_selected_files = cast(Optional[list[int]], metainfo.pop(b"selected_files", None))

--- a/src/tribler/core/restapi/file_endpoint.py
+++ b/src/tribler/core/restapi/file_endpoint.py
@@ -51,13 +51,13 @@ class FileEndpoint(RESTEndpoint):
                                  "paths": paths})
 
         # Move up until we find a directory
-        path = Path(path).resolve()
-        while not path.is_dir():
-            path = path.parent
+        parent_path = Path(path).resolve()
+        while not parent_path.is_dir():
+            parent_path = parent_path.parent
 
         # Get all files/subdirs
         results = []
-        for file in path.iterdir():
+        for file in parent_path.iterdir():
             if not file.is_dir() and not show_files:
                 continue
             with contextlib.suppress(PermissionError):
@@ -68,11 +68,11 @@ class FileEndpoint(RESTEndpoint):
         # Get parent path (if available)
         results.insert(0, {
             "name": "..",
-            "path": str(path.parent.resolve()) if path != path.parent else "/",
+            "path": str(parent_path.parent.resolve()) if parent_path != parent_path.parent else "/",
             "dir": True,
         })
 
-        return RESTResponse({"current": str(path.resolve()),
+        return RESTResponse({"current": str(parent_path.resolve()),
                              "paths": results})
 
     async def list(self, request: web.Request) -> RESTResponse:

--- a/src/tribler/core/restapi/rest_manager.py
+++ b/src/tribler/core/restapi/rest_manager.py
@@ -261,7 +261,7 @@ class RESTManager:
                                    str(e))
             raise
 
-        current_port = api_port or self.site._server.sockets[0].getsockname()[1]  # noqa: SLF001
+        current_port = api_port or cast(Server, self.site._server).sockets[0].getsockname()[1]  # noqa: SLF001
         self.config.set("api/http_port_running", current_port)
         self.config.write()
 
@@ -280,7 +280,7 @@ class RESTManager:
         await self.site_https.start()
         self._logger.info("Started HTTPS REST API: %s", self.site_https.name)
 
-        current_port = port or self.site_https._server.sockets[0].getsockname()[1]  # noqa: SLF001
+        current_port = port or cast(Server, self.site_https._server).sockets[0].getsockname()[1]  # noqa: SLF001
         self.config.set("api/https_port_running", current_port)
         self.config.write()
 

--- a/src/tribler/core/restapi/webui_endpoint.py
+++ b/src/tribler/core/restapi/webui_endpoint.py
@@ -1,3 +1,5 @@
+from __future__ import annotations
+
 import logging
 import mimetypes
 
@@ -20,25 +22,30 @@ class WebUIEndpoint(RESTEndpoint):
         """
         super().__init__()
         self._logger = logging.getLogger(self.__class__.__name__)
-        self.app.add_routes([web.get('/{path:.*}', self.return_files)])
+        self.app.add_routes([web.get("/{path:.*}", self.return_files)])
 
         self.webui_root = tribler.get_webui_root()
-        self.has_dist = (self.webui_root / 'dist').exists()
+        self.has_dist = (self.webui_root / "dist").exists()
         self.session = ClientSession() if not self.has_dist else None
 
-    async def return_files(self, request: web.Request) -> RESTResponse:
+    async def return_files(self, request: web.Request) -> RESTResponse | web.FileResponse:
         """
         Return the file at the requested path.
         """
-        path = request.match_info['path'] or 'index.html'
+        path = request.match_info["path"] or "index.html"
 
         if self.session:
-            async with self.session.get(f'http://localhost:5173/{path}') as response:
-                return web.Response(body=await response.read(), content_type=response.content_type)
+            async with self.session.get(f"http://localhost:5173/{path}") as client_response:
+                return RESTResponse(body=await client_response.read(), content_type=client_response.content_type)
         else:
-            resource = self.webui_root / 'dist' / path
+            resource = self.webui_root / "dist" / path
             response = web.FileResponse(resource)
-            response.content_type = 'application/javascript' if path.endswith('.tsx') else mimetypes.guess_type(path)[0]
+            if path.endswith(".tsx"):
+                response.content_type = "application/javascript"
+            elif (guessed_type := mimetypes.guess_type(path)[0]) is not None:
+                response.content_type = guessed_type
+            else:
+                response.content_type = "application/octet-stream"
             return response
 
     async def shutdown_task_manager(self) -> None:

--- a/src/tribler/core/socks5/aiohttp_connector.py
+++ b/src/tribler/core/socks5/aiohttp_connector.py
@@ -2,12 +2,15 @@ from __future__ import annotations
 
 import socket
 from asyncio import BaseTransport, wait_for
-from typing import Callable
+from typing import TYPE_CHECKING, Callable
 
 from aiohttp import TCPConnector
 from aiohttp.abc import AbstractResolver
 
 from tribler.core.socks5.client import Socks5Client, Socks5ClientUDPConnection
+
+if TYPE_CHECKING:
+    from aiohttp.abc import ResolveResult
 
 
 class FakeResolver(AbstractResolver):
@@ -15,7 +18,7 @@ class FakeResolver(AbstractResolver):
     Pretend to resolve an address. Just echo it back.
     """
 
-    async def resolve(self, host: str, port: int = 0, family: int = socket.AF_INET) -> list[dict[str, str | int]]:
+    async def resolve(self, host: str, port: int = 0, family: int = socket.AF_INET) -> list[ResolveResult]:
         """
         Resolve a host to itself.
         """

